packages/vector_graphics_compiler/lib/src/svg/masking_optimizer.dart

```diff
@@ -0,0 +1,362 @@
+class MaskingOptimizer extends Visitor<_Result, Node>
+    with ErrorOnUnResolvedNode<_Result, Node> {
+  /// List of masks to add.
+  final List<ResolvedPathNode> masksToApply = <ResolvedPathNode>[];
+
+  /// Applies visitor to given node.
+  Node apply(Node node) {
+    final Node newNode = node.accept(this, null).node;
+    return newNode;
+  }
+
+  /// Applies mask to a path node, and returns resulting path node.
+  ResolvedPathNode applyMask(Node child, ResolvedPathNode maskPathNode) {
+    final ResolvedPathNode pathNode = child as ResolvedPathNode;
+    final path_ops.Path maskPathOpsPath = toPathOpsPath(maskPathNode.path);
+    final path_ops.Path pathPathOpsPath = toPathOpsPath(pathNode.path);
+    final path_ops.Path intersection =
+        maskPathOpsPath.applyOp(pathPathOpsPath, path_ops.PathOp.intersect);
+    final Path newPath = toVectorGraphicsPath(intersection);
+    final ResolvedPathNode newPathNode = ResolvedPathNode(
+        paint: pathNode.paint, bounds: maskPathNode.bounds, path: newPath);
+
+    maskPathOpsPath.dispose();
+    pathPathOpsPath.dispose();
+    intersection.dispose();
+
+    return newPathNode;
+  }
+
+  @override
+  _Result visitEmptyNode(Node node, void data) {
+    final _Result _result = _Result(node);
+    return _result;
+  }
+
+  /// Visits applies optimizer to all children of ResolvedMaskNode.
+  _Result visitChildren(Node node, _Result data) {
+    if (node is ResolvedMaskNode) {
+      data = node.child.accept(this, data);
+    }
+    return data;
+  }
+
+  @override
+  _Result visitParentNode(ParentNode parentNode, Node data) {
+    final List<Node> newChildren = <Node>[];
+    bool deleteMaskNode = true;
+
+    for (Node child in parentNode.children) {
+      final _Result childResult = child.accept(this, parentNode);
+      newChildren.add(childResult.node);
+      if (childResult.deleteMaskNode == false) {
+        deleteMaskNode = false;
+      }
+    }
+
+    final ParentNode newParentNode = ParentNode(parentNode.attributes,
+        precalculatedTransform: parentNode.transform, children: newChildren);
+
+    final _Result _result = _Result(newParentNode);
+
+    _result.deleteMaskNode = deleteMaskNode;
+
+    return _result;
+  }
+
+  @override
+  _Result visitMaskNode(MaskNode maskNode, Node data) {
+    final _Result _result = _Result(maskNode);
+
+    return _result;
+  }
+
+  @override
+  _Result visitPathNode(PathNode pathNode, Node data) {
+    final _Result _result = _Result(pathNode);
+    return _result;
+  }
+
+  @override
+  _Result visitResolvedMaskNode(ResolvedMaskNode maskNode, void data) {
+    _Result _result = _Result(maskNode);
+    final ResolvedPathNode? singleMaskPathNode = getSingleChild(maskNode.mask);
+
+    if (singleMaskPathNode != null) {
+      masksToApply.add(singleMaskPathNode);
+      final _Result childResult = maskNode.child.accept(this, maskNode);
+      masksToApply.removeLast();
+
+      if (childResult.deleteMaskNode == true) {
```

Comment:
This can't ever be null so you can use it as is

```suggestion
      if (childResult.deleteMaskNode) {
```